core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala

```diff
@@ -101,13 +101,29 @@ class KryoSerializer(conf: SparkConf)
     }
   }
 
-  @transient
-  var pool: KryoPool = getPool
+  private class PoolWrapper extends KryoPool {
+    private var pool: KryoPool = getPool
+
+    override def borrow(): Kryo = pool.borrow()
+
+    override def release(kryo: Kryo): Unit = pool.release(kryo)
+
+    override def run[T](kryoCallback: KryoCallback[T]): T = pool.run(kryoCallback)
+
+    def reset(): Unit = {
+      pool = getPool
```

Comment:
OK. We could look at updating to Kryo 5 in Spark 3, too, if there were other benefits.
Does the old pool need to get freed or anything here?

Reply:
The KryoPool interface exposes no way to free it, so I believe that is not necessary

Reply:
OK, and it only gets recreated when a new classloader is set, it seems, so it is rare.